model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/singleUseObject/entity/JpaSingleUseObjectEntity.java

```diff
@@ -0,0 +1,225 @@
+    @Version
+    @Column
+    private int version;
+
+    @Type(type = "jsonb")
+    @Column(columnDefinition = "jsonb")
+    private final JpaSingleUseObjectMetadata metadata;
+
+    @Column(insertable = false, updatable = false)
+    @Basic(fetch = FetchType.LAZY)
+    private Integer entityVersion;
+
+    @Column(insertable = false, updatable = false)
+    @Basic(fetch = FetchType.LAZY)
+    private String objectKey;
+
+    @Column(insertable = false, updatable = false)
+    @Basic(fetch = FetchType.LAZY)
+    private Long expiration;
+
+    @OneToMany(mappedBy = "root", cascade = CascadeType.PERSIST, orphanRemoval = true)
+    private final Set<JpaSingleUseObjectNoteEntity> notes = new HashSet<>();
+
+    /**
+     * No-argument constructor, used by hibernate to instantiate entities.
+     */
+    public JpaSingleUseObjectEntity() {
+        this.metadata = new JpaSingleUseObjectMetadata();
+    }
+
+    public JpaSingleUseObjectEntity(final DeepCloner cloner) {
+        this.metadata = new JpaSingleUseObjectMetadata(cloner);
+    }
+
+    public boolean isMetadataInitialized() {
+        return this.metadata != null;
+    }
+
+    @Override
+    public int getVersion() {
+        return this.version;
+    }
+
+    @Override
+    public Integer getEntityVersion() {
+        if (this.isMetadataInitialized()) return this.metadata.getEntityVersion();
+        return this.entityVersion;
+    }
+
+    @Override
+    public void setEntityVersion(Integer entityVersion) {
+        this.metadata.setEntityVersion(entityVersion);
+    }
+
+    @Override
+    public String getId() {
+        return id == null ? null : id.toString();
+    }
+
+    @Override
+    public void setId(String id) {
+        String validatedId = UuidValidator.validateAndConvert(id);
+        this.id = UUID.fromString(validatedId);
+    }
+
+    @Override
+    public String getObjectKey() {
+        if (this.isMetadataInitialized()) return this.metadata.getObjectKey();
+        return this.objectKey;
+    }
+
+    @Override
+    public void setObjectKey(final String objectKey) {
+        this.metadata.setObjectKey(objectKey);
+    }
+
+    @Override
+    public Integer getCurrentSchemaVersion() {
+        return CURRENT_SCHEMA_VERSION_SINGLE_USE_OBJECT;
+    }
+
+    @Override
+    public String getActionId() {
+        return this.metadata.getActionId();
+    }
+
+    @Override
+    public void setActionId(String actionId) {
+        this.metadata.setActionId(actionId);
+    }
+
+    @Override
+    public String getActionVerificationNonce() {
+        return this.metadata.getActionVerificationNonce();
+    }
+
+    @Override
+    public void setActionVerificationNonce(String actionVerificationNonce) {
+        this.metadata.setActionVerificationNonce(actionVerificationNonce);
+    }
+
+    @Override
+    public Map<String, String> getNotes() {
+        // avoid using Collectors::toMap as it throws NPE if the note value is unset (null).
```

Comment:
Does that mean we need to store `null`s in note values?

note: https://github.com/keycloak/keycloak/issues/9741

Reply:
> perhaps the right way to fix it would be to change the code that adds the notes so it doesn't add a note if it has no value.

+1 for this solution

Reply:
+1 to fixing the code where `null` is put, but not in this PR. Still, we need to be defensive here, so even if there is `null` value then the code needs to work.

Think of no-downtime upgradability - if one version supported `null` and the other did not, the code must not fail with NPE.

Explanation and workaround is summarized e.g. in https://stackoverflow.com/a/24632080/6930869

Reply:
I'll leave this code as is then, and open a separate tracker to fix the inclusion of a null value - need to find the test that was failing.

Reply:
Please paste a link to the separate issue here once it's created, thank you.

Reply:
I've tracked the failing tests and the `null` values are usually either the `scope` or `nonce` in `OAuth2DeviceCodeModel`. Opened https://github.com/keycloak/keycloak/issues/12774.

I was thinking about being defensive - wouldn't it be better if we didn't store null-valued notes in the first place? In other words, adding `null` checks in `setNote()` and `setNotes` to discard non-null notes. Then `getNotes` could safely use `Collectors::toMap` knowing that non-null values do not exist in the store.

Reply:
Hmm, looking into `setNotes` method implementation:
```
@SuppressWarnings("unchecked") @Override public void setNotes(java.util.Map<String,String> p0) {
        p0 = p0 == null ? null : new java.util.HashMap<>(p0);
        if (p0 != null) {
            p0.values().removeIf(org.keycloak.models.map.common.UndefinedValuesUtils::isUndefined);
            if (org.keycloak.models.map.common.UndefinedValuesUtils.isUndefined(p0)) p0 = null;
        }
        updated |= ! Objects.equals(fNotes, p0);
        fNotes = p0;
    }
```
this should already remove undefined values.

Reply:
For the record: I've used exactly it in user/cllient sessions: https://github.com/keycloak/keycloak/pull/12241/files#diff-f8c7267d81dd0ea60117e905f425105d651dda638b403d2c06e80b3eed5c5cceR302

Reply:
I'll follow the same path then, and have single-use object notes consistent with the user session notes.